### PR TITLE
Support Project (Protected) Environments (if license allows it)

### LINKED ISSUE #444: Environments support
From what I gathered currently there is no way for provider to create Project Environments nor there is a way to manage Protected Environments. Seems that at least normal Environments are supported by github.com/xanzy/go-gitlab

Any plans to add a resource for it?

### PR DESCRIPTION
**Pull request by sirlatrom**
WIP on Protected Environments.

Adds resource `gitlab_project_protected_environment`.

I've added `gitlab_project_environment` as well, with a rather minimal set of attributes.

Fixes #444.

### CODE REVIEW

**Comment by sirlatrom** [comment]
Ping @armsnyder for opinions

**Comment by sirlatrom** [comment]
Oops, forgot to wait for https://github.com/xanzy/go-gitlab/pull/1066 before marking as ready 😅

File: gitlab/resource_gitlab_project_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_environment.go:41]: Can you add this attribute to the doc page

File: gitlab/resource_gitlab_project_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_environment.go:66]: resp always needs a nil check, since the err could be anything.

File: gitlab/resource_gitlab_project_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_environment.go:125]: Usually use HasChange in the update function

```suggestion
	if d.HasChange("external_url") {
		options.ExternalURL = gitlab.String(d.Get("external_url").(string))
```

File: gitlab/resource_gitlab_project_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_environment.go:160]: Is the StopEnvironment API call idempotent? For example, if the DeleteEnvironment call fails and a user retries the delete, will the StopEnvironment API call fail on the second try? (It shouldn't.)

File: gitlab/resource_gitlab_project_environment_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_environment_test.go:171]: Is this unfinished, or can we delete this empty if block?

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_protected_environment.go:63]: The schema validation could be improved here.

1. `user_id` and `group_id` should probably use integer validations. Since you don't use these attributes in your acceptance test, there's a good chance the `validation.StringIsNotEmpty` validation has a bug here.

2. The API docs say that exactly one of `access_level`, `user_id`, and `group_id` must be specified. You can check for that easily using `ExactlyOneOf`.

3. Usually for enum attributes like `access_level` we validate against the list of valid access levels (see https://github.com/gitlabhq/terraform-provider-gitlab/blob/9884585ddf50efa061798e452c1e3375b8714af7/gitlab/resource_gitlab_project_membership.go#L42)

File: gitlab/resource_gitlab_project_protected_environment_test.go
Comment (armsnyder) [inline gitlab/resource_gitlab_project_protected_environment_test.go:236]: Can you add test steps for:
1. Having more than once `deploy_access_levels` blocks
2. Using `group_id` and `user_id`.

**Review by armsnyder — CHANGES_REQUESTED** [review]

File: gitlab/resource_gitlab_project_environment.go
Comment (sirlatrom) [reply gitlab/resource_gitlab_project_environment.go:160]: Stop:

> It returns 200 if the environment was successfully stopped, and 404 if the environment does not exist. 

https://docs.gitlab.com/ee/api/environments.html#stop-an-environment

Delete:

> It returns 204 if the environment was successfully deleted, and 404 if the environment does not exist.

https://docs.gitlab.com/ee/api/environments.html#delete-an-environment

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (sirlatrom) [reply gitlab/resource_gitlab_project_protected_environment.go]: Regarding point 2, when trying to set `ExactlyOneOf`, there is no way to achieve that if multiple blocks are expected to be supported:

>```
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>ExactlyOneOf: []string{"deploy_access_levels.0.acces_level", "deploy_access_levels.0.user_id", "deploy_access_levels.0.group_id"},
>```

File: gitlab/resource_gitlab_project_environment.go
Comment (sirlatrom) [reply gitlab/resource_gitlab_project_environment.go:41]: Done in be501eb73c189081777f409af7cc2dd3b5ccc77a

File: gitlab/resource_gitlab_project_protected_environment_test.go
Comment (sirlatrom) [reply gitlab/resource_gitlab_project_protected_environment_test.go:249]: Done in 4d9f680437b362b3a2e3071a9e0a1a362fc58672

**Comment by sirlatrom** [comment]
The upstream PR has been merged, so now I'm only waiting on the next release before I can mark this as ready.

**Comment by nicholasklick** [comment]
@sirlatrom looks like https://github.com/xanzy/go-gitlab/pull/1066 is merged.

Can we merge this MR now? Any blockers?

**Comment by nicholasklick** [comment]
Any other concerns from @armsnyder?

File: gitlab/resource_gitlab_project_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_environment.go:92]: Can we add a nil check for `resp` here?

```suggestion
		if resp != nil && resp.StatusCode == http.StatusNotFound {
```

File: gitlab/resource_gitlab_project_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_environment.go:103]: I know it's currently done this way all over the provider, but it would be good to handle `error` for `Set`:

```suggestion
	err = d.Set("project", project)
	if err != nil {
		return err
	}

	err = d.Set("name", environment.Name)
	if err != nil {
		return err
	}

	err = d.Set("state", environment.State)
	if err != nil {
		return err
	}
```

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment.go:112]: Would be good to handle `error` for `Set`:

```suggestion
	err = d.Set("project", project)
	if err != nil {
		return err
	}

	err = d.Set("environment", environment)
	if err != nil {
		return err
	}
```

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment.go:128]:
```suggestion
	err = d.Set("environment", protectedEnvironment.Name)
	if err != nil {
		return err
	}
```

File: gitlab/resource_gitlab_project_environment_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_environment_test.go:180]:
```suggestion
	if response != nil && response.StatusCode != 404 {
```

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment.go:93]:
```suggestion
		if resp != nil && resp.StatusCode == http.StatusNotFound {
```

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment.go:120]:
```suggestion
		if resp != nil && resp.StatusCode == http.StatusNotFound {
```

File: gitlab/resource_gitlab_project_protected_environment_test.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment_test.go:179]:
```suggestion
	if response != nil && response.StatusCode != 404 {
```

**Review by mattkasa — CHANGES_REQUESTED** [review]
@sirlatrom this looks really good! I left some suggestions, all error handling for `Set` or checking `resp != nil` before using `resp` as @armsnyder suggested, if those are ok with you I'd be happy to merge 👍

File: gitlab/resource_gitlab_project_environment.go
Comment (armsnyder) [reply gitlab/resource_gitlab_project_environment.go:103]: I'm gonna disagree here, as skipping error checking with primitive types is a regular practice in TF providers, since the SDK will catch these issues for us. Guidance for non-primitive types here: https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types

Comment (sirlatrom) [reply gitlab/resource_gitlab_project_environment.go:103]: @armsnyder I suppose that goes for the rest of the suggested error checks, too?

Comment (armsnyder) [reply gitlab/resource_gitlab_project_environment.go:103]: Yeah, but maybe wait until we come to some consensus in #651 😄

**Comment by sirlatrom** [comment]
Sorry if my rebase on the master branch has broken the suggested code changes. I can see there's an ongoing discussion about the error checking of `d.Set(...)`. Until that is resolved, couldn't this PR follow the existing approach of not error checking when the argument is a primitive type, and then be handled by #651 after the fact?

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (mattkasa) [inline gitlab/resource_gitlab_project_protected_environment.go:215]: This seems to be duplicating `accessLevelNameToValue` (https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/util.go#L15), but with some differences (is `noone` correct, or `no one`?).

Would it make sense to fix/update `accessLevelNameToValue` instead?

**Comment by mattkasa** [comment]
> couldn't this PR follow the existing approach of not error checking when the argument is a primitive type, and then be handled by #651 after the fact?

Absolutely 👍

**Review by mattkasa — COMMENTED** [review]
This looks really good, just one question about `accessLevelStringToValue`.

File: gitlab/resource_gitlab_project_protected_environment.go
Comment (sirlatrom) [reply gitlab/resource_gitlab_project_protected_environment.go:215]: Yes, certainly. I'll make the change, remove the above code and replace uses of it with `accessLevelNameToValue`.

File: docs/resources/project_protected_environment.md
Comment (mattkasa) [inline docs/resources/project_protected_environment.md:51]: > Valid values are `developer` or `maintainer`.

Sorry if I'm missing it, where is this being validated?

Comment (sirlatrom) [reply docs/resources/project_protected_environment.md:51]: Only server-side, I don't think I've added validation in the schema.

Comment (sirlatrom) [reply docs/resources/project_protected_environment.md:51]: I guess the note on the API docs means I can use StringInSlice for validating `developer`, `maintainer` or `admin`: https://docs.gitlab.com/ee/api/protected_environments.html#valid-access-levels

Comment (sirlatrom) [reply docs/resources/project_protected_environment.md:51]: Turns out I did validate it here, though it would also accept "reporter", etc.

https://github.com/gitlabhq/terraform-provider-gitlab/blob/2cc27a4e0f881ad64f590f58df7290f98c69f575/gitlab/resource_gitlab_project_protected_environment.go#L45-L50

Would you like me to change it to strictly allow `developer`, `maintainer` or `admin`? The UI only provides the first two as options.

Comment (mattkasa) [reply docs/resources/project_protected_environment.md:51]: @sirlatrom Yeah, I think it would be good to make the `ValidateFunc` match the documentation 👍

Comment (sirlatrom) [reply docs/resources/project_protected_environment.md:51]: That should be done now, and I've rebased on master, too.

**Comment by campers** [comment]
@sirlatrom thanks for your work on this, was there any more work that needed to be done on this PR other than fixing the lint error?

**Comment by sirlatrom** [comment]
> @sirlatrom thanks for your work on this, was there any more work that needed to be done on this PR other than fixing the lint error?

I cannot recall any other outstanding issues, but I've since moved on to another employer and will not be able to spend any more time on this.

**Comment by tahir-gvs** [comment]
Any chance of this getting released soon?

**Pull request closed without merging**
